rules/manager.go

```diff
@@ -523,6 +527,91 @@ func (m *Manager) Update(interval time.Duration, files []string) error {
 	return nil
 }
 
+// RestoreForState restores the 'for' state of the alerts
+// by looking up last ActiveAt from storage.
+func (m *Manager) RestoreForState(ts time.Time) {
+
+	maxtMS := int64(model.TimeFromUnixNano(ts.UnixNano()))
+	for _, newg := range m.groups {
+		for _, rule := range newg.Rules() {
+			alertRule, ok := rule.(*AlertingRule)
+			if !ok {
+				continue
+			}
+			_, err := alertRule.Eval(m.opts.Context, ts, newg.opts.QueryFunc, newg.opts.ExternalURL)
+			if err != nil {
+				level.Error(m.logger).Log("msg", "Failed to restore 'for' state",
+					labels.AlertName, alertRule.Name(), "stage", "Eval", "err", err)
+				continue
+			}
+
+			mint := ts.Add(-m.opts.OutageTolerance)
+			mintMS := int64(model.TimeFromUnixNano(mint.UnixNano()))
+
+			q, err := m.opts.TSDB.Querier(m.opts.Context, mintMS, maxtMS)
+			if err != nil {
+				level.Error(m.logger).Log("msg", "Failed to restore 'for' state",
+					labels.AlertName, alertRule.Name(), "stage", "Querier", "err", err)
+				continue
+			}
+
+			alertHoldDuration := alertRule.HoldDuration()
+
+			alertRule.ForEachActiveAlert(func(a *Alert) {
+
+				smpl := alertRule.forStateSample(a, time.Now(), 0)
+				var matchers []*labels.Matcher
+				for _, l := range smpl.Metric {
+					mt, _ := labels.NewMatcher(labels.MatchEqual, l.Name, l.Value)
+					matchers = append(matchers, mt)
+				}
+
+				sset, err := q.Select(nil, matchers...)
+				if err != nil {
+					level.Error(m.logger).Log("msg", "Failed to restore 'for' state",
+						labels.AlertName, alertRule.Name(), "stage", "Select", "err", err)
+					return
+				}
+
+				seriesFound := false
+				var s storage.Series
+				for sset.Next() {
+					if len(sset.At().Labels()) == len(smpl.Metric) {
+						s = sset.At()
+						seriesFound = true
+						break
+					}
+				}
+
+				if !seriesFound {
+					return
+				}
+
+				// Series found for the 'for' state.
+				var v float64
+				it := s.Iterator()
+				for it.Next() {
+					_, v = it.At()
+				}
+				if value.IsStaleNaN(v) { // Alert was not active.
+					return
+				}
+
+				restoredTime := time.Unix(int64(v), 0)
+				if alertHoldDuration > m.opts.ForGracePeriod && restoredTime.Add(alertHoldDuration).Sub(ts) < m.opts.ForGracePeriod {
+					restoredTime = ts.Add(m.opts.ForGracePeriod).Add(-alertHoldDuration)
```

Comment:
I don't think this is right, it always waits another 10m. If I've a 2h alert that I'm only 5m into and do an instant restart, I should have to wait the remaining 1h55m - not only 10m.

Reply:
`restoredTime.Add(alertHoldDuration).Sub(ts) < m.opts.ForGracePeriod` takes care of it.
`restoredTime.Add(alertHoldDuration).Sub(ts)` will be `1h55m`, which is > grace period. So it wont go inside the `if` block.

To explain the logic more clearly:
If alert became active at `0` and prometheus was down for `5m`, then:
1. `restoredTime` will be 0.
2. `alertHoldDuration > m.opts.ForGracePeriod`: we see that the hold duration is greater than grace period to apply the logic of grace period.
3. `restoredTime.Add(alertHoldDuration).Sub(ts)` can be viewed as

```
// ts = current time
  restoredTime                        ts                      firing wrt restoredTime
       |-------------------------------|-------------------------------|

       ^-----------------------alertHoldDuration-----------------------^

                                       ^-------------------------------^
                                     restoredTime.Add(alertHoldDuration).Sub(ts)

       ^--------------5m--------------^^-------------1h55m-------------^
```

4. `ts.Add(m.opts.ForGracePeriod).Add(-alertHoldDuration)`: it makes the alert to fire at `ts.Add(m.opts.ForGracePeriod)`

As `restoredTime.Add(alertHoldDuration).Sub(ts)` (`1h55m`) > `m.opts.ForGracePeriod` (`10m`), the `if` block wont run.

Reply:
It'll never go inside the if for an alert with more than a 10m for - which can't be right.

Reply:
Isn't it the other way round?
`alertHoldDuration > m.opts.ForGracePeriod` specifies that it goes inside `if` only when the `for` is `>10m`. If `for` was `<10`, then pending the alert for another `10m` won't be right, hence doesn't go inside in that case.

Reply:
> it's discarding the timestamp of the sample

The timestamp of the sample comes to use indirectly when querying. As we query for `ALERTS_FOR_STATE` in the outage tolerance period, the timestamp of the samples come into picture in the query processing.

If you talk about explicitly using the timestamp of the sample, its not needed, as the `ActiveAt` is the value of the sample.

I have updated the comments to make the logic more clear. Please let me know which part doesn't seem fine.